File: Data Structures & Algorithms/longest-increasing-subsequence/submission-5.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    lengthOfLIS(nums) {
        let n = nums.length;
        const d = new Map();
        function dfs(index){
            if(index < 0)
              return ;
            let val = nums[index];
            if(d.has(val)){
                let i = index+1;
                while(val !== nums[i]){
                    if(val < nums[i])
                      d.set(val, Math.max(d.get(val),d.get(nums[i])+1));
                    i++;
                }
            }
            else{
                let count =0;
                for(let i=index+1;i<n;i++){
                    if(val < nums[i])
                      count = Math.max(count, d.get(nums[i]));
                }
                d.set(val,count+1);
            }
            dfs(index-1);
        }
        dfs(n-1);
        let m=0;
        for(let key of d.keys()){
            if(m<d.get(key))
              m=d.get(key);
        }
        return m;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-5.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        n,m=len(s1),len(s2)
        t=len(s3)
        d={}
        if t != n+m :
            return False
        def dfs(i,j,k):
            if (i,j,k) in d :
                return  d[(i,j,k)]
            if k==t:
                if i==n and j==m:
                    return True
                return False
            if i<n and s3[k]==s1[i] :
                if dfs(i+1,j,k+1):
                    d[(i,j,k)]=True
                    return True
            if j<m and s3[k]==s2[j]:
                if dfs(i,j+1,k+1) :
                    d[(i,j,k)]=True
                    return True
            d[(i,j,k)]=False
            return False
        return dfs(0,0,0)